File: app/src/main/java/me/tatarka/betterfragment/sample/dagger/FragmentModule.java
package me.tatarka.betterfragment.sample.dagger;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;
import me.tatarka.betterfragment.Fragment;
import me.tatarka.betterfragment.sample.MyFragment;
import me.tatarka.betterfragment.sample.NavigationFragment;
import me.tatarka.betterfragment.sample.SimpleHostFragment;
import me.tatarka.betterfragment.sample.ViewPagerFragment;

@Module
public abstract class FragmentModule {

    @Binds
    abstract Fragment.Factory factory(DaggerFragmentFactory factory);

    @Binds
    @IntoMap
    @FragmentKey(MyFragment.class)
    abstract Fragment myFragment(MyFragment fragment);

    @Binds
    @IntoMap
    @FragmentKey(SimpleHostFragment.class)
    abstract Fragment simpleHostFragment(SimpleHostFragment fragment);

    @Binds
    @IntoMap
    @FragmentKey(NavigationFragment.class)
    abstract Fragment navigationFragment(NavigationFragment fragment);

    @Binds
    @IntoMap
    @FragmentKey(ViewPagerFragment.class)
    abstract Fragment viewPagerFragment(ViewPagerFragment fragment);
}

File: app/src/main/java/me/tatarka/betterfragment/sample/dagger/FragmentModule.kt
package me.tatarka.betterfragment.sample.dagger

import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap
import me.tatarka.betterfragment.app.Fragment
import me.tatarka.betterfragment.sample.*

@Module
abstract class FragmentModule {

    @Binds
    internal abstract fun factory(factory: DaggerFragmentFactory): Fragment.Factory

    @Binds
    @IntoMap
    @FragmentKey(MyFragment::class)
    internal abstract fun myFragment(fragment: MyFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(SimpleHostFragment::class)
    internal abstract fun simpleHostFragment(fragment: SimpleHostFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(NavigationFragment::class)
    internal abstract fun navigationFragment(fragment: NavigationFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(ViewPagerFragment::class)
    internal abstract fun viewPagerFragment(fragment: ViewPagerFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(DialogHostFragment::class)
    internal abstract fun dialogHostFragment(fragment: DialogHostFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(MyAlertDialogFragment::class)
    internal abstract fun myAlertDialogFragment(fragment: MyAlertDialogFragment): Fragment

    @Binds
    @IntoMap
    @FragmentKey(SimpleDialogFragment::class)
    internal abstract fun simpleDialogFragment(fragment: SimpleDialogFragment): Fragment
}
